python/packages/autogen-ext/src/autogen_ext/agents/azure/_azure_ai_agent.py

```diff
@@ -0,0 +1,867 @@
+import dotenv
+from autogen_agentchat import EVENT_LOGGER_NAME
+from autogen_agentchat.agents import BaseChatAgent
+from autogen_agentchat.base import Response
+from autogen_agentchat.messages import (
+    AgentEvent,
+    ChatMessage,
+    HandoffMessage,
+    MultiModalMessage,
+    StopMessage,
+    TextMessage,
+    ToolCallExecutionEvent,
+    ToolCallRequestEvent,
+)
+from autogen_core import CancellationToken, FunctionCall
+from autogen_core.models._model_client import ChatCompletionClient
+from autogen_core.models._types import FunctionExecutionResult
+from autogen_core.tools import FunctionTool, Tool
+from pydantic import BaseModel, Field
+
+import azure.ai.projects.models as models
+from azure.ai.projects import _types
+from azure.ai.projects.aio import AIProjectClient
+from azure.identity.aio import DefaultAzureCredential
+
+event_logger = logging.getLogger(EVENT_LOGGER_NAME)
+
+
+class AzureAIAgentState(BaseModel):
+    """
+    Represents the state of an AzureAIAgent that can be saved and loaded.
+
+    This state model keeps track of persistent information about an agent session
+    including agent and thread identifiers, message history, and associated resources.
+
+    Attributes:
+        type (str): The type identifier for the state object, always "AzureAIAgentState"
+        agent_id (Optional[str]): The ID of the Azure AI agent
+        thread_id (Optional[str]): The ID of the conversation thread
+        initial_message_ids (List[str]): List of message IDs from the initial state
+        vector_store_id (Optional[str]): The ID of the associated vector store for file search
+        uploaded_file_ids (List[str]): List of IDs for files uploaded to the agent
+    """
+
+    type: str = Field(default="AzureAIAgentState")
+    agent_id: Optional[str] = None
+    thread_id: Optional[str] = None
+    initial_message_ids: List[str] = Field(default_factory=list)
+    vector_store_id: Optional[str] = None
+    uploaded_file_ids: List[str] = Field(default_factory=list)
+
+
+class AzureAIAgent(BaseChatAgent):
+    """
+    Azure AI Assistant agent for AutoGen.
+
+    This agent leverages the Azure AI Assistant API to create AI assistants with capabilities like:
+
+    * Code interpretation and execution
+    * Grounding with Bing search
+    * File handling and search
+    * Custom function calling
+    * Multi-turn conversations
+
+    The agent integrates with AutoGen's messaging system, providing a seamless way to use Azure AI
+    capabilities within the AutoGen framework. It supports tools like code interpreter,
+    file search, and various grounding mechanisms.
+
+    Agent name must be a valid Python identifier:
+        1. It must start with a letter (A-Z, a-z) or an underscore (_).
+        2. It can only contain letters, digits (0-9), or underscores.
+        3. It cannot be a Python keyword.
+        4. It cannot contain spaces or special characters.
+        5. It cannot start with a digit.
+
+    Examples:
```

Comment:
I think it is useful to move the notebook examples into the API docs, we do that for other extensions. 

For each example in the API doc we make sure it is runnable as a script, and type checked using `pyright`.